Fixed #35535 -- Add simple block tag (#18343)
django/django: docs/howto/custom-template-tags.txt

    {% current_time "%Y-%m-%d %I:%M %p" as the_time %}
    <p>The time is {{ the_time }}.</p>

.. _howto-custom-template-tags-simple-block-tags:

Simple block tags
-----------------

.. method:: django.template.Library.simple_block_tag()

> Passing arguments may not be enough for some usages -- you may need to pass chunks of
> rendered template into your custom tags. This can be achieved using the
> ``simple_block_tag`` helper function of ``django.template.Library``.
> ``simple_block_tag`` functions similar to :meth:`~django.template.Library.simple_tag()`,
> however additionally passes a ``content`` argument which contains its rendered template
> content.

Reviewer: ```suggestion
If you need to pass a section of rendered template into your custom tag, this
can be achieved using the ``simple_block_tag`` helper function of
``django.template.Library``. ``simple_block_tag`` functions similar to
:meth:`~django.template.Library.simple_tag()`, but also passes a ``content``
argument which contains the rendered template content block contained by the
tag.
```
So I guess with this, the content itself is an argument and other arguments are not required (if that makes sense)

For example, a custom block tag which created a chart could look like this::

    from django import template
    from myapp.charts import render_chart

    register = template.Library()

    @register.simple_block_tag
    def chart(content):

>         return render_chart(source=content)

Author: Something which occured to me, is that the use-cases of a tag like this are fairly limited. However, a variant which accepts a template and processes context (akin to `inclusion_tag`) could be incredibly helpful. Most of the building blocks are here to make that happen, but not sure if that warrants a separate PR, or it's worth trying to squeeze in here? I'm not sure what the diff size would be, which might be the deciding factor.

Author: I've tried this out, and the diff is quite small. `InclusionNode` can be subclassed in much the same way as was done with `SimpleNode`, with similar duplication of `inclusion_tag`. 

With that said, there's little overlap between the 2 feature, so it probably does make sense they be separate. I can contribute those once this is merged.

> The ``content`` argument contains everything in between the ``{% chart %}`` and
> ``{% endchart %}`` tags. If there are other template tags or variables, they will be
> rendered before being passed in.

Reviewer: ```suggestion
The ``content`` argument contains everything in between the ``{% chart %}`` and
``{% endchart %}`` tags. If there are other template tags or variables, they
will be rendered before being passed in.
```
Also in other places, docs need some wrapping adjustments

.. code-block:: html+django

    {% chart %}
        digraph G {
            label = "Chart for {{ request.user }}"
            A -> {B C}
        }
    {% endchart %}

A few things to note about the ``simple_block_tag`` helper function:

* The first argument must be called ``content``, and will contain the contents of the
  template tag as a rendered string.
* Checking for the required number of arguments, etc., has already been
  done by the time our function is called, so we don't need to do that.
* The quotes around the argument (if any) have already been stripped away,
  so we receive a plain string.

> * If the argument was a template variable, our function is passed the
>   current value of the variable, not the variable itself.

Reviewer: ```suggestion
* If an argument was a template variable, our function is passed the current
  value of the variable, not the variable itself.
```
As we already have the argument "content", all other arguments are essentially additional and optional I think

    register.simple_block_tag(lambda content: content.upper(), name="uppercase")

    @register.simple_block_tag(name="lowercase")
    def some_function(content):
        return content.lower()

``simple_block_tag`` functions may accept any number of positional or keyword
arguments. For example::

    @register.simple_block_tag
    def mytag(content, a, b, *args, **kwargs):
        profile = kwargs["profile"]
        ...
        return ...

Then in the template any number of arguments, separated by spaces, may be
passed to the template tag. Like in Python, the values for keyword arguments
are set using the equal sign ("``=``") and must be provided after the
positional arguments. For example:

.. code-block:: html+django

    {% mytag 123 "abcd" book.title profile=user.profile %}
        This is some extra content.
    {% endmytag%}

It's possible to store the tag results in a template variable rather than
directly outputting it. This is done by using the ``as`` argument followed by
the variable name. Doing so enables you to output the content yourself where
you see fit:

.. code-block:: html+django

    {% chart as my_chart %}
        A -> {B C}
    {% endchart %}

>     <p>The chart is {{ the_time|length }} characters.</p>

Reviewer: ```suggestion
    <p>The chart is {{ my_chart|length }} characters.</p>
```

    {% current_time "%Y-%m-%d %I:%M %p" as the_time %}
    <p>The time is {{ the_time }}.</p>

.. _howto-custom-template-tags-simple-block-tags:

Simple block tags
-----------------

.. versionadded:: 5.2

.. method:: django.template.Library.simple_block_tag()

> If you need to pass a section of rendered template into your custom tag, the
> ``simple_block_tag`` helper function of ``django.template.Library`` functions
> similar to :meth:`~django.template.Library.simple_tag()`, but also passes a
> ``content`` argument which contains its rendered template content contained by
> the tag.

Reviewer: ```suggestion
If you need to pass a section of rendered template into your custom tag, the
``simple_block_tag`` helper function of ``django.template.Library`` passes a
``content`` argument, containing the rendered template content enclosed in the
tag. This functions similar to :meth:`~django.template.Library.simple_tag()`,
but with an additional ``content`` argument.
```
This sentence is quite long. I wonder if we can split it a bit

* The first argument must be called ``content``, and will contain the contents
  of the template tag as a rendered string.
* Checking for the required number of arguments, etc., has already been
  done by the time our function is called, so we don't need to do that.
* The quotes around the argument (if any) have already been stripped away,
  so we receive a plain string.
* If an argument was a template variable, our function is passed the current
  value of the variable, not the variable itself.
* Variables passed to the tag are not included in the rendering context of the
  content, as would be using the ``{% with %}`` tag.

Like ``simple_tag``, ``simple_block_tag`` passes its output through
:func:`~django.utils.html.conditional_escape` if the template context is in
autoescape mode, to ensure correct HTML and protect you from XSS
vulnerabilities.

If additional escaping is not desired, you will need to use
:func:`~django.utils.safestring.mark_safe` if you are absolutely sure that your
code does not contain XSS vulnerabilities. For building small HTML snippets,
use of :func:`~django.utils.html.format_html` instead of ``mark_safe()`` is

> strongly recommended.
> 
> Because the ``content`` argument has already been rendered by Django, it is
> already escaped. Therefore, it may be necessary to use
> :func:`~django.utils.safestring.mark_safe` to ensure it's not double-escaped.
> 
> If your template tag needs to access the current context, you can use the
> ``takes_context`` argument when registering your tag::
> 
>     @register.simple_block_tag(takes_context=True)
>     def chart(context, content):
>         return render_chart(source=content, dark_mode=context["dark_mode"])
> 
> Note that the first argument *must* be called ``context``, and ``content``
> moves to the second argument.
> 
> For more information on how the ``takes_context`` option works, see the section
> on :ref:`inclusion tags<howto-custom-template-tags-inclusion-tags>`.
> 
> If you need to rename your tag, you can provide a custom name for it::
> 
>     register.simple_block_tag(lambda content: content.upper(), name="uppercase")
> 
> 
>     @register.simple_block_tag(name="lowercase")
>     def some_function(content):
>         return content.lower()
> 
> ``simple_block_tag`` functions may accept any number of positional or keyword
> arguments. For example::
> 
>     @register.simple_block_tag
>     def mytag(content, a, b, *args, **kwargs):
>         profile = kwargs["profile"]
>         ...
>         return ...
> 
> Then in the template any number of arguments, separated by spaces, may be
> passed to the template tag. Like in Python, the values for keyword arguments
> are set using the equal sign ("``=``") and must be provided after the
> positional arguments. For example:
> 
> .. code-block:: html+django
> 
>     {% mytag 123 "abcd" book.title profile=user.profile %}
>         This is some extra content.
>     {% endmytag%}
> 
> It's possible to store the tag results in a template variable rather than
> directly outputting it. This is done by using the ``as`` argument followed by
> the variable name. Doing so enables you to output the content yourself where
> you see fit:
> 
> .. code-block:: html+django
> 
>     {% chart as my_chart %}
>         A -> {B C}
>     {% endchart %}
>     <p>The chart is {{ my_chart|length }} characters.</p>
> 

Reviewer 2: I think that there are too many bits that are duplicated from the "simple tags" explanation. I think those should be framed under a summary, helping reducing the duplication. How about something like:

```suggestion
* Variables passed to the tag are not included in the rendering context of the
  content, as would be when using the ``{% with %}`` tag.

Just like :ref:`simple_tag<howto-custom-template-tags-simple-tags>`,
``simple_block_tag``:

* Validates the quantity and quality of the arguments.
* Quotes are stripped from arguments if necessary.
* Output is escaped accordingly.
* Context can be accessed by passing ``takes_context=True`` at registration
  time. Note that in this case, the first argument *must* be called
  ``context``, and ``content`` must follows.
* The tag can be renamed using passing the ``name`` argument when registering.
* ``simple_block_tag`` functions may accept any number of positional or keyword
  arguments.
* The tag result can be stored in a template variable using the ``as`` variant.

.. admonition::  Content Escaping

    ``simple_block_tag`` behaves similarly to ``simple_tag`` regarding
    auto-escaping. For details on escaping and safety, refer to ``simple_tag``.

    Because the ``content`` argument has already been rendered by Django, it is
    already escaped. Therefore, it may be necessary to use
    :func:`~django.utils.safestring.mark_safe` to ensure it's not
    double-escaped.
```

    already escaped.

A more complete example
~~~~~~~~~~~~~~~~~~~~~~~

Suppose a Django app, ``testapp``, needs to define a general message box to be
used in templates, supporting various message levels and content beyond a
simple phrase. The tag could be used as follows:

.. code-block:: python
    :caption: ``testapp/templatetags/testapptags.py``

    from django import template
    from django.utils.html import format_html

    register = template.Library()

    @register.simple_block_tag(takes_context=True)
    def msgbox(context, content, level):
        format_kwargs = {
            "level": level.lower(),
            "level_title": level.capitalize(),
            "content": content,
            "open": " open" if level.lower() == "error" else "",
            "site": context.get("site", "My Site"),
        }
        result = """
        <div class="msgbox {level}">
          <details{open}>
            <summary>
              <strong>{level_title}</strong>: Please read for <i>{site}</i>
            </summary>
            <p>
              {content}
            </p>
          </details>
        </div>
        """

>         return format_html(result, **format_kwargs)
> 
> .. code-block:: python
>     :caption: ``testapp/views.py``
> 
>     from django.shortcuts import render
> 
> 
>     def simpleblocktag_view(request):
>         return render(request, "test.html", context={"site": "Important Site"})
> 
> 
> Then, ``test.html`` would look like this:
> 
> .. code-block:: html+django
>     :caption: ``testapp/templates/test.html``
> 
>     {% extends "base.html" %}
> 
>     {% load testapptags %}
> 
>     {% block content %}
> 
>       {% msgbox level="error" %}
>         Please fix all errors. Further documentation can be found at
>         <a href="http://example.com">Docs</a>.
>       {% endmsgbox %}
> 
>       {% msgbox level="info" %}
>         More information at: <a href="http://othersite.com">Other Site</a>/
>       {% endmsgbox %}
> 
>     {% endblock %}
> 
> With a resulting HTML as follows:
> 
> .. code-block:: html
> 
>     <div class="msgbox error">
>       <details open>
>         <summary>
>           <strong>Error</strong>: Please read for <i>Important Site</i>
>         </summary>
>         <p>
>           Please fix all errors. Further documentation can be found at
>           <a href="http://example.com">Docs</a>.
>         </p>
>       </details>
>     </div>
> 
>     <div class="msgbox info">
>       <details>
>         <summary>
>           <strong>Info</strong>: Please read for <i>Important Site</i>
>         </summary>
>         <p>
>           More information at: <a href="http://othersite.com">Other Site</a>
>         </p>
>       </details>
>     </div>

Reviewer: I like that the example shows how this is useful in a more real way, I think it's too complete and I wouldn't put it in it's own section with a header. I would propose:
```suggestion
For a more complete example, a message box that supports different message
levels and content beyond a simple phrase could be implemented via a tag as
follows::

    from django import template
    from django.utils.html import format_html

    register = template.Library()

    @register.simple_block_tag()
    def msgbox(context, content, level, title):
        format_kwargs = {
            "level": level,
            "level_title": level.capitalize(),
            "content": content,
            "open": " open" if level == "error" else "",
            "title": title,
        }
        result = """
        <details{open} class="msgbox {level}">
          <summary>
            <strong>{level_title}!</strong>: {title}</i>
          </summary>
            <p>
              {content}
            </p>
          </details>
        """
        return format_html(result, **format_kwargs)

Then, the following template snippet:

.. code-block:: html+django

    {% msgbox level="error" title="You did something wrong!" %}
      Please fix all errors. Further documentation can be found at
      <a href="http://example.com">Docs</a>.
    {% endmsgbox %}

    {% msgbox level="info" title="Here is some help" %}
      More information at: <a href="http://othersite.com">Other Site</a>/
    {% endmsgbox %}

Would result in HTML as follows:

.. code-block:: html

    <details open class="msgbox error">
      <summary>
        <strong>Error!</strong>: You did something wrong!
      </summary>
      <p>
         Please fix all errors. Further documentation can be found at
         <a href="http://example.com">Docs</a>.
      </p>
    </details>

    <details open class="msgbox error">
      <summary>
        <strong>Info!</strong>: Here is some help
      </summary>
      <p>
        More information at: <a href="http://othersite.com">Other Site</a>
      </p>
    </details>

.. admonition::  Content Escaping

    ``simple_block_tag`` behaves similarly to ``simple_tag`` regarding
    auto-escaping. For details on escaping and safety, refer to ``simple_tag``.
    Because the ``content`` argument has already been rendered by Django, it is
    already escaped.
```

Reviewer 2: Thank you @sarahboyce ! Following our conversation where we realized this is part of howto docs (with a mix of reference docs which prompted the initial comment), I pushed some wording tweaks to the example but otherwise left it as a more hand-holding description.

* Variables passed to the tag are not included in the rendering context of the
  content, as would be when using the ``{% with %}`` tag.

Just like :ref:`simple_tag<howto-custom-template-tags-simple-tags>`,
``simple_block_tag``:

* Validates the quantity and quality of the arguments.
* Strips quotes from arguments if necessary.
* Escapes the output accordingly.
* Supports passing ``takes_context=True`` at registration time to access context.
  Note that in this case, the first argument *must* be called ``context``,
  and ``content`` must follows.
* Supports renaming the tag by passing the ``name`` argument when registering.
* Supports accepting any number of positional or keyword arguments.
* Supports storing the result in a template variable using the ``as`` variant.

.. admonition::  Content Escaping

    ``simple_block_tag`` behaves similarly to ``simple_tag`` regarding
    auto-escaping. For details on escaping and safety, refer to ``simple_tag``.

    Because the ``content`` argument has already been rendered by Django, it is
    already escaped. Therefore, it may be necessary to use
    :func:`~django.utils.safestring.mark_safe` to ensure it's not

>     double-escaped.

Reviewer 2: I think this is missing an empty line after the paragraphs ends:
```suggestion
    double-escaped.

```

* Variables passed to the tag are not included in the rendering context of the
  content, as would be when using the ``{% with %}`` tag.

Just like :ref:`simple_tag<howto-custom-template-tags-simple-tags>`,
``simple_block_tag``:

* Validates the quantity and quality of the arguments.
* Strips quotes from arguments if necessary.
* Escapes the output accordingly.
* Supports passing ``takes_context=True`` at registration time to access context.
  Note that in this case, the first argument *must* be called ``context``,
  and ``content`` must follows.
* Supports renaming the tag by passing the ``name`` argument when registering.
* Supports accepting any number of positional or keyword arguments.
* Supports storing the result in a template variable using the ``as`` variant.

.. admonition::  Content Escaping

    ``simple_block_tag`` behaves similarly to ``simple_tag`` regarding
    auto-escaping. For details on escaping and safety, refer to ``simple_tag``.

    Because the ``content`` argument has already been rendered by Django, it is
    already escaped. Therefore, it may be necessary to use
    :func:`~django.utils.safestring.mark_safe` to ensure it's not

>     double-escaped.

Reviewer 3: Does this refer to the following from the `simple_tag` docs?

> Unlike other tag utilities, simple_tag passes its output through conditional_escape() if the template context is in autoescape mode, to ensure correct HTML and protect you from XSS vulnerabilities.
> If additional escaping is not desired, you will need to use mark_safe() if you are absolutely sure that your code does not contain XSS vulnerabilities. For building small HTML snippets, use of format_html() instead of mark_safe() is strongly recommended.

If so, I think it would be worth replacing the sentence above with something similar, because it's much less ambiguous. 

Otherwise, at least I'd suggest using only `format_html` instead of `mark_safe`.

Also, adding some html markup in the inner tag contents to all tests that rely on returning `content` would also help clarify what needs to be escaped and how exactly.

Author: No, this is slightly different. The `simple_tag` docs are specifically about the _output_ of the tag, which is still true for `simple_block_tag`.

This point is about the provided `content` string, which is different.

Reviewer 3: So, for example if I add some html markup in the block contents then `end_one_param_block` doesn't render it as html tags because the interpolation is done incorrectly in the tag definition.

But `escape_format_html_block` still works because it uses `format_html`, which calls `conditional_escape` on `content` and returns a safe string. If I do `return content` as is that works too.

These all seems expected to me, so I'm not sure what this note refers to exactly.

(Also, if there's html in the contents then that's NOT "escaped" in `content` but `content` is a safe string that renders as html, so I don't think that's a correct statement either.)

Reviewer 3: I might be still missing something but I think even the whole admonition could be removed because I don't see anything surprising about how this works. The `content` argument is a `SafeString` (as I'd expect it to be, but that might be still worth pointing out) and all the standard django rules about handling those apply here too.

Reviewer 2: @balazs-endresz Thanks. I have pushed a few changes where I add a more complete example that shows how `format_html` is used. Let me know what you think!

Reviewer 3: Thank you, the docs change and the examples look good to me.
